### DISCUSSION TITLE
Add a basic, standalone xaml-free Hello World example to the docs

### DISCUSSION THREAD

**Discussion by tseguine** [AvaloniaUI/Avalonia#9006] [Ideas]
According to the FAQ it is possible to use Avalonia without xaml, but I could find almost zero information in the official documentation for how to accomplish this. Maybe I just didn't find it, but it seems like if it exists it is well hidden.

It is really frustrating from a user perspective when the only introductory way of "Getting Started" in the docs is "Hey, install this visual studio extension and download this entire project that does 1000 things you don't want or need."

I understand it is more or less the intended way of using the framework, but I don't think it is too out of line to somewhere document the 3 liner to get a basic window to display:

`Application app = Application.Current ?? AppBuilder.Configure<Application>().UsePlatformDetect().SetupWithoutStarting().Instance;`
`app.Styles.Add(new DefaultTheme());`
`app.Run(new Window() { Title = "Avalonia Basic Example", Content = "Hello Avalonia!" });`

As someone with no prior Avalonia, WPF, or Winforms experience it took me several hours to figure out just those three lines. And I somehow think I am not the only person whose use case fits this way of using the library better than the xaml path.

**Comment by kekekeks** [comment]
The _relevant_ docs about using Avalonia without XAML are here: https://docs.avaloniaui.net/docs/data-binding/binding-from-code

Those don't include basic stuff like constructing an object graph since we assume that you know what are you doing.

**Reply by alexzandros** [reply]
Being rude does not make you right.

**Reply by StefanH-AT** [reply]
Well you just confirmed my suspicions. Avalonia is one of the hardest to learn frameworks I have ever had to use. It's somehow the best option yet the docs seem like they're designed to actively discouraging anyone from understanding your software. "We assume we know what you're doing". Have you considered people go to the docs to learn?

**Reply by thevortexcloud** [reply]
> yet the docs seem like they're designed to actively discouraging anyone from understanding your software. 

Honestly, I would say you should look at WPF examples. Don't get me wrong, I also wish Avalonia had better docs. But WPF docs are much more plentiful and about 60-80% of it will work in Avalonia with only minor changes. It's also far easier to find WPF tutorials.

EG the page for how to use Bindings in Avalonia looks like this:

https://docs.avaloniaui.net/docs/basics/data/data-binding/

Compared to WPF's:

https://learn.microsoft.com/en-us/dotnet/desktop/wpf/data/?view=netdesktop-9.0

As you can see, WPF's is _significantly_ more detailed. How you use binding is also basically identical between the two as well (the biggest difference is Avalonia has compiled and reflection bindings, while WPF only has reflection bindings).

Once you learn one XAML framework, you have basically learned all of them.

**Reply by StefanH-AT** [reply]
I actually came to avalonia with some basic wpf experience. Thanks for the resources! My main troubles with avalonia just come from the seemingly enforced axaml and mvvm pattern. After using some javascript frameworks like react and vuejs, this overly declarative way of writing applications just seems overcomplicated. I may see the benefit of these patterns for really, really large applications. What I like about avalonia is the cross-platform rendering and OS integrations. This is really good. I think avalonia has a lot of potential so it just sucks to see it so unwelcoming to new users. Just staring at the docs it doesn't look like there's a way to use avalonia without a full axaml mvvm reactiveui setup since that's all that's ever shown. So now I'm trying to explore options of using avalonia without axaml just for its core multi-platform rendering capabilities

**Reply by Trigve** [reply]
I don't know what your prior experience is with WPF is (you've said you have some basic one), but I was struggling with the Avalonia too as first, but then I've bought the `WPF 4.5 Unleashed` (even if this book is a bit outdated) book and after reading even the first few chapters a lot of things were clear to me. I do think that Avalonia assume you have the prior WPF knowledge (hence the Avalonia name).

**Reply by thevortexcloud** [reply]
> look like there's a way to use avalonia without a full axaml mvvm reactiveui setup since that's all that's ever shown

There are in fact docs for not using ReactiveUI for Avalonia. They are just not easy to find, and a lot of them are not on the docs site. EG there is the entire sample repo.

https://github.com/AvaloniaUI/Avalonia.Samples

You can also use CommunityMVVM if you don't feel like implementing everything yourself. But a lot of that stems from the fact that Avalonia historically only supported ReactiveUI out of the box, and nobody has written newer docs for it since.

As for MVVM: While you can make apps without it, I would strongly suggest you don't. Avalonia views are often stateless. Meaning as soon as you can't see them any more, you lose all the data in them. The View Model is what holds the state. This is also true for WPF. It also makes binding significantly harder if you don't use MVVM. Which is also probably why you wont find many guides on how to do it.

> avalonia without axaml just for its core multi-platform rendering capabilities

If you just want to avoid XAML, there is an official library to make it easier.

https://github.com/AvaloniaUI/Avalonia.Markup.Declarative

I personally prefer XAML though, since it ensures a clean separation between view and non view logic.

**Reply by StefanH-AT** [reply]
This is great! Pretty much exactly what I wanted thanks! I'll give this a go after work

**Reply by stevensrmiller** [reply]
Have a look at my Github repo. I've posted several Avalonia examples with no XAML at all: https://github.com/stevensrmiller/AvaloniaWithoutXAML

**Reply by thevortexcloud** [reply]
> but then I've bought the WPF 4.5 Unleashed (even if this book is a bit outdated) book

There was someone working on an Avalonia book. Not sure where they got to with it though.

https://github.com/AvaloniaUI/Avalonia/discussions/14798#discussioncomment-10536671

**Reply by stevemonaco** [reply]
IMO, you should use the best supported approach for the tool. MVVM and XAML exist for good reasons that have been tested and proven successful for nearly 20 years.

Likewise, I wouldn't be complaining that JS frontends, Flutter, etc don't support MVVM or XAML. Especially in webapps where you already have strong decoupling -- most apps are thin clients that don't do much relative to a thick client. The data comes neatly packaged from an API call and you push visuals to a very stable browser that handles the entire layout, virtualization, rendering, etc for you.

Even if you push beyond this, eventually you will need to learn XAML. Every theming library author uses XAML for their `ControlTheme`s, so if you want to customize, you will eventually need to read their XAML so you can properly troubleshoot.

That said, https://github.com/IvanJosipovic/KubeUI is a good example of a XAML-free app. It still uses MVVM. You likely won't be able to avoid MVVM in any sane way to me. Reproducing bindings via observables or writing manual synchronization code in code-behind is not something that I would find reasonable.

**Reply by lucasmontec** [reply]
IMO, this is a reason not to use Avalonia. Or at least, to remove the quote that says it supports XAML-free UI development.
Most use cases don't need "successful" 20-year, web-compatible, multi-platform, layout-separated, fat frameworks. C# doesn't seem to have a good UI framework, and Avalonia is what comes up on Google searches. People asking for this just want to be able to have some level of UI without having to touch disgraceful, old, and dirty XML. So maybe consider adding useful information instead of personal opinions that just sound condescending.

**Reply by stevemonaco** [reply]
@lucasmontec My personal opinions are from experience, both in work and being a part of the larger .NET community for a long time. If you want to be a trailblazer and potentially shoot yourself in the foot many times, I gave a path and discussed the tradeoffs. This is how advice and perspective works. If you're That Guy with additional weeks or months to burn, then feel free to disregard my advice, work towards building that ecosystem, and convince people that it's a better approach through demonstrated successful projects. That's simply not the reality today and I'm not going to misrepresent it. The vast majority of beginners would rather spend time writing their app using the pit of success and full C# approaches are not for beginners because you need to know what you're doing moreso than XAML. And that's something that can't be understood until you have specific experience.

Avalonia is definitely a good UI framework for the audience that wants long lifecycle, crossplatform, native apps with deeply customizable theming. That might not be the right fit for you currently, but that's ok.

**Reply by lucasmontec** [reply]
@stevemonaco 
" My personal opinions are from experience," yeah, completely disregarding others' experiences and objectives. Even those who have more experience than you in areas in which you have none.
"both in work and being a part of the larger .NET community for a long **time**" Just another fallacy of authority trying to keep up with the condescending behavior. I'm also a professional and a part of the community. So are those asking for OSS to advertise what it actually supports. So are those that have wildly different objectives using tools like this. Objectives and directions that you seem to disregard because you don't understand or don't care. Most probably because of your long years doing the same things on and on.
" I gave a path" -> "you should use the best supported approach for the tool" 
"discussed the tradeoffs" -> "MVVM and XAML exist for good reasons" tradeoffs = reasons. Then you just went on rambling about JS and complaining about web tech. I fail to see ANY tradeoffs in your comment. Just frustrations.
"and convince people that it's a better approach." It's funny how blind you are to scopes outside of your realm, and yet try to sell your silver bullet approach to software development as if you have found some sort of holy grail of a pattern. The fact that you measure the extra time it would take in months shows how little you understand the community. Some projects could be done in a matter of hours. XAML will indeed add days and even weeks to those 1-2h GUI apps. This attitude is what makes people shy away from C# and choose other languages. Not that this is a bad thing! But we need to be better. C# be more than this. 
I'm not talking about beginners here, I'm talking about people with at least 10+ years of professional software dev in their backs (me included). 
The dude literally just asked where to find docs about something that is literally in the landing page of the system, and somehow this made everyone go crazy on what they should and shouldn't do. Think about how crazy this is!

**Reply by thevortexcloud** [reply]
>  yeah, completely disregarding others' experiences and objectives. Even those who have more experience than you in areas in which you have none.
>  Just another fallacy of authority trying to keep up with the condescending behavior. 

Did you even read what you wrote?

**Reply by stevemonaco** [reply]
I'm actively providing advice, even if it's unclear. If you have questions about why, ask them instead of berating me. If you aren't interested in productive conversation, please take your 10+ years of professional experience elsewhere and bother them on a Friday night instead.

"I gave a path" -> KubeUI. Other projects, that I'd normally mention, were already mentioned above.

"discussed tradeoffs" -> You still need to read XAML because everybody else uses, including themes. Needing to work with observables instead of bindings for compile-time (if you want to avoid reflection, at the moment).

```
It's funny how blind you are to scopes outside of your realm, and yet try to sell your silver bullet approach to software development as if you have found some sort of holy grail of a pattern. The fact that you measure the extra time it would take in months shows how little you understand the community.
```

It's not a silver bullet, but XAML with MVVM is demonstrably the most successful approach in open source and amongst the professional .NET developers I've spoken with. In some frameworks (namely UWP), using XAML is hugely more performant over the equivalent C#.

I don't mean months of full-time work, I mean months of tweaking here and there to find the good patterns in the C# side and across the visual hierarchy that you will now be manually maintaining. Then a new C# version comes out and you need to tweak it again. Modern .NET UI frameworks have a large learning curve and you will be iterating on ways to do things for quite awhile on your own simply because the amount of pure C# apps for any major GUI framework is barren.

"XAML will indeed add days and even weeks to those 1-2h GUI apps" -> Have you used XAML for longer than a tutorial app? It's clearly not the case for anybody who uses a XAML framework in their professional work. 95%+ of the learning curve is elsewhere (layout, binding, navigation, styling, etc) though XAML tends to take the blame.

"The dude literally just asked where to find docs about something that is literally in the landing page of the system, and somehow this made everyone go crazy on what they should and shouldn't do."

Yes and they were linked to docs relevant to bindings. Told that they needed to learn how to use an object graph. Which is a fancy term for using `new` and assigning child properties which is 90%+ of written XAML.

The docs accept community contributions. I've contributed a few small C#-related snippets for things that are typically done in XAML. While there probably should be a bare minimum tutorial with just C#, there's relatively zero serious interest. However, there are a lot of zealous advocates for pure C# / declarative UI approaches and yet very few of them contribute or make open source projects (in the .NET world). Which makes me wonder if any of them would ever cross the finish line when making something that isn't a throwaway weekend project.

The overall challenge here is: 1. writing two sets of examples for everything in the docs. 2. choosing to spend time doing this instead of writing about still undocumented features.

**Reply by MikeCodesDotNET** [reply]
> I'm also a professional

Then I ask that you start acting like one.

>  and a part of the community.

Coming into our GitHub discussions with rudeness doesn’t make you a community member. It makes you look entitled and unpleasant.

> So are those asking for OSS to advertise what it actually supports.

Avalonia apps can be built with C#. That’s a fact. Just because you don’t know how doesn’t mean it can’t be done.

 > because of your long years doing the same things on and on.

Unnecessary, arrogant, and entirely unacceptable in our community.

> shows how little you understand the community.

This says more about you than about @stevemonaco. Steve is one of the most active and supportive people in this ecosystem. Suggesting he doesn’t understand our community is, frankly, absurd.

Your tone and attitude are not welcome here. I am leaving you unblocked for now, but I expect your behaviour to change. If it doesn’t, you will be blocked from our org.

**Reply by jm-abellan** [reply]
I can't really understand why anybody would talk like that about @stevemonaco, who is the very first one answering any question in the Avalonia channels
You may not like what he says. Then, just don't follow his suggestions
Personally, I have always found very helpful his answers

**Reply by stevensrmiller** [reply]
My goodness. This issue seems to bring out some passions.

Anyone who needs some self-contained, running Avalonia without XAML might like my examples at [https://github.com/stevensrmiller/AvaloniaWithoutXAML](https://github.com/stevensrmiller/AvaloniaWithoutXAML). They mostly target NET8.0, but can easily be set to NET9.0. I've been using them for a couple of years now with my students who have some, but usually not very much, prior programming experience. The students seem to get the idea of creating objects, so creating a Window, a Canvas, and several Line objects is no harder with or without XAML. The Window contains the Canvas. The Canvas contains the Line objects. That's not too hard either.

I'm sure there are lots of situations where XAML provides advantages, so I don't argue for or against it. But you don't need it to use Avalonia. For what I'm doing, Avalonia without XAML is working very well. I'm happy to answer questions about my samples, if anyone has any.

**Reply by lucasmontec** [reply]
Yeah, I misbehaved, but reading this thread and the other answers made me do that. Still, not justified.
It seems that in .NET community, this entitled behavior disregarding people's actual needs and motivations is quite common. This is maddening. 

Also, it seems that there's no current cross-platform solution for GUI in .NET where people can spin real small or targeted UI solutions without having to deal with XAML or other data languages targeting layout structure. This causes people to find tools like Avalonia and causes them to ask about the **advertised** feature's documentation. So it is natural that once the reaction becomes: "you're wrong because I have more experience, go learn and be better", people (including me) get mad.

The use case for XAML you listed is still far away from what I need from any UI framework. Even the issues you noted, I knew about those, and they were all non-issues for the solutions I need from such a framework.

But I did overreact, and I understand that this doesn't have a place, neither here nor anywhere in OSS or outside of it.

> Then I ask that you start acting like one.

 Stuff like this is also crazy. I don't work for you, nor do I have to be professional here. I'm professional towards those who are professional with me or towards those who aren't, but pay me.

> Coming into our GitHub discussions with rudeness

 I merely answered a comment that was already very rude. Might not have been in your view, but it was in mine. Still, I agree that I could have de-escalated instead.

> Suggesting he doesn’t understand our community 

I agree, but I wasn't talking about your community. OSS is bigger than you, and this is why there are many people here asking you guys why you'd say you support something and yet answer like it's something wrong.

But yeah, I read what I wrote and I did mess up badly and escalated this a LOT. Sorry for that. Still, I'd kindly ask you guys to try to understand the perspective of others and maybe ask clarification questions on why they want something before you reach any conclusions.

Now, about use cases for C# only UI: C# only allows for quick prototyping. Small 1-screen internal tools with UI that can be written in an hour and honestly never need to be updated. Small tools for engineering (GUI for configuration and debugging) to accompany hardware products that will never be client-facing. Game launcher screens. Self-Updated apps that need to just display, maybe a progress bar, and a settings panel. There is a very large subspace of projects that would benefit from a plain code API. 
Still, I'd say that Avalonia is not the tool for those, but right now, in C#, there aren't many options. When searching, people will find Avalonia and Uno, both stating support for code-only interfaces and both providing near-zero documentation for it.

**Reply by stevemonaco** [reply]
> Small tools for engineering (GUI for configuration and debugging) to accompany hardware products that will never be client-facing. Game launcher screens. Self-Updated apps that need to just display, maybe a progress bar, and a settings panel.

Things like this can already be somewhat complicated in terms of project structure (multiple source files for UI components), with or without XAML and MVVM. There are external projects like NXUI that demonstrate that apps with a truly small scope may be reasonable for some. With .NET 10, you can run C# scripts even:

https://github.com/wieslawsoltes/NXUI?tab=readme-ov-file#dotnet-run-appcs

I personally can't recommend using foundational projects that aren't yet proven to scale well (scope creep is a thing) or may go unmaintained. Especially one that would likely force a total app rewrite. The situation is unlike a third-party UI component going unmaintained -- those have more options to be replaced with limited impact.

**Reply by thevortexcloud** [reply]
It sounds like you want something like ImGUI. 

https://github.com/ImGuiNET/ImGui.NET/blob/master/src/ImGui.NET.SampleProgram/Program.cs

Granted: ImGUI is not a native C# UI framework. It is however a very popular C/C++ one for creating quick GUIs and is often used for things like debug menus in games.

**Reply by lucasmontec** [reply]
Yeah, ImGUI would be sufficient, but it's not optimal. The window management in ImGUI is very bad. It was made to create debug and inspection tools for real-time applications, so whenever you run it not targeting an external window, it creates a weird window system with some transparent overlays that cover the whole screen to be able to render its setup.

Python offers some stuff, but it looks pretty bad, and it simply doesn't have the power C# has for the operations side.
Java has old and new UI systems, but Java is much better when run using a web server. Local Java is sub-optimal, and system interaction is somewhat limited compared with the C# api.

For C#, for now, I'm using Uno for my personal client-facing projects (auto-updaters, game/software launchers, and small game dev apps - content managers and such), and EtoForms for my apps that interface with hardware (hotplate control system, custom input hardware controllers/configurators, and such).
For experimental apps, I still go for Python.

Most of those are very small scopes, with well-defined boundaries. They don't scale past a few UI classes, usually 2 to 5 classes with at most 1 or 2 screens. Even better, because they are easy to write and easy to discard. Real throwaways. Either they work and are considered done, or they scale and get rewritten with an appropriate framework or are just trashed after the concept is tested.

There's a studio building an alternative. Sirenix is building a C# native, SDF-based immediate mode rendering UI system, but it's pre-alpha and doesn't show in any searches.

**Reply by MikeCodesDotNET** [reply]
I have blocked Lucas from the org due to repeated disrespectful communication. 

While his points about the lack of documentation around XAML-less Avalonia are valid, the way they were expressed was not.

**Reply by stevensrmiller** [reply]
This forum item seems to attract a badly behaved commenter every so often. The quality of the existing Avalonia documentation is quite good, so it sets a high bar for new additions to it. Regardless, do you think my AvaloniaWithoutXAML samples could be refactored in some way so as to serve as, at least, introductory documentation? I want Avalonia to be popular with all potential users. It's actually _very_ easy to use it without XAML. It's just not obvious how you do it.

Let me know if you think this idea has potential.

**Comment by tseguine** [comment]
None of that stuff is actually pertinent or in any way helpful to the basic task of starting an application and  creating a window. i.e. **The first thing anyone would actually want to do with the library**

**Comment by tseguine** [comment]
I also think you misunderstood my gripe: WPF which this API is apparently based on does not have this documentation issue. The architecture is slightly different so it makes sense that it is simpler, but in WPF it is both:

- Obvious enough to figure out in 5 seconds without any documentation whatsoever by a simple glance at the class names.
- **Is actually documented anyway** in an easily discoverable location in the docs.

It is a FAQ. That means you know this is something people have asked how to do. And even coming from WPF it wouldn't exactly be obvious how to do it in Avalonia.

I am merely suggesting giving an answer in the FAQ that is actually useful to someone who is actually trying to do the frequently asked thing.

My steps:

- Do the obvious thing, that would have actually worked in WPF.
- Fail because the platform layer is not initialized.
- Search for how to instantiate the platform layer.
- Find literally nothing except for some obscure posts on github where someone with the same issue getting the fairly unhelpful comment from you "You need to follow the framework initialization procedure, see project templates for examples." (you literally could have told them how to do what they were trying to do with less typed characters than what you actually told thim)
- Finally found a real actual answer that isn't "hey: download this example project. "
- No error, but window is inexplicably blank.
- Search for a reason
- find nothing
- actually download the template in frustration
- that works even though it is doing almost the exact same thing as I am line for line except loading xaml files.
- Go back to the example that was actually helpful
- Figure out you have to manually load the default theme for anything to actually display whatsoever.

Every single step is in my opinion reasonable and logical. I have used **several** different GUI libraries in **several** different languages in over 20 years of programming. And it has literally never in my life (even when I was entirely new to programming and a literal child) taken me more than 15 minutes to get the goddamn hello world application running without 5 files and over 100 lines of code.

Rant over. It is a usability issue. Period. Fix it or don't. At least if someone else has the same problem as me now, they will find this post instead of your unhelpful attitude of "we assume that you know what are you doing." GTFO with that bull.

**Comment by grokys** [comment]
Hi sorry about that, we're trying our best but we're trying as hard as we can. We don't have a team the size of some of the frameworks you may be used to using, currently and so documentation can be lacking for some use-cases (particularly if you're trying to avoid XAML). We'll try to add something to the docs about this.

**Comment by sn4k3** [comment]
> I have used several different GUI libraries in several different languages in over 20 years of programming...

I only have used WinForms before, it took me few days to adapt and learn the framework but almost all questions and troubles were solved by look at documentation, which is not perfect but is enough to start, more when you have that such experience...

If you use Visual Studio with Avalonia extension, it's easy as create new project and select Avalonia. It will create a empty "Hello World" app for you with all requirements and things setup for you, no different from create a WinForm project.
Also look here: https://docs.avaloniaui.net/docs/getting-started to see step by step how to work with Avalonia.

About your title question, once by random I found an library or sample that spawn a window from code without any xaml and extra files, since I was not looking for it I forgot where I saw that or the name, but then it's possible...

Someone here could give you an anwser to obtain that, but your atitude isn't the best and may lead to people skip on you. 
Ok documentation aren't complete and far from perfect, but look what this team did for the community of C# and desktop applications, for free and open-source! So far you don't have any C# alternative to multi-platform that behaves like Avalonia. Even MAUI is "far" and will not follow the same principle as Avalonia.
You could have just ask your question and then fork and update docs. Learn, share and retribute what others give to you.

**Reply by tseguine** [reply]
@sn4k3 
> Someone here could give you an anwser to obtain that, but your atitude isn't the best

My attitude was completely normal until someone told me "we assume you know what you are doing" when I came with a reasonable gripe. Following precisely the tone that I was approached with.

And at this point I basically already figured it out myself. My point was that I shouldn't have had to do that for something so basic.

**Comment by MikeCodesDotNET** [comment]
Hi @tseguine 

I'm sorry that you've found several challenges with starting development with Avalonia, and I appreciate you taking the time to let us know. 

We're aware that our documentation has gaps, and it's something we're keen to resolve. Our docs have been written from the perspective that developers will be writing their apps using XAML, reflecting how most of our community uses Avalonia. 

Given our open-source nature, the most helpful thing you could do for us (other than contribute some docs) would be to create issues in the documentation repository to let us know of missing content. 

Lastly, if the docs haven't helped, I suggest asking in our telegram community chat. That chat has hundreds of active Avalonia developers with a wealth of experience who can help.

**Comment by tseguine** [comment]
@MikeCodesDotNET 

Are the docs also in github, and are you open to pull requests? If so (and I find time: using this is for a personal project not work) I might try to write something up.

**Reply by MikeCodesDotNET** [reply]
https://github.com/AvaloniaUI/Documentation

**Reply by timunie** [reply]
Hey @tseguine ,

Yes the docs are open source and PR are welcome. Additionally we have Avalonia.Samples which is also open source and should grow over time. 

You can reach out to me via gitter or telegram and we can prepare a hello world sample together, which can then be linked to the docs. 

Welcome to Avalonia and happy coding
Tim

**Comment by kkukshtel** [comment]
As a note I also ran into this exact issue — I read on the first page of the documentation that you could use it without XAML and then searched everywhere for literally _any_ documentation on how to do it and the only thing I found was this Github issue. It's been more than two years since this was first reported, is there a canonical C# code-only example?

**Reply by thevortexcloud** [reply]
There is now a dedicated library to make it easier with some small examples : https://github.com/AvaloniaUI/Avalonia.Markup.Declarative

**Reply by tseguine** [reply]
I actually didn't intend to leave this hanging. I figured out how to do what I set out to do, and I was going to make a PR on the docs, but never got around to it because of life. It actually isn't that hard, there are just more gotchas than one would suspect.

If you want as minimal as it gets, then the three lines I posted in my original post basically do it. What djeikyb posted just recently is more or less what I original used (I dunno if it was that exact code but it was similar) as a working example to figure out what the minimal example is.

You are still going to run into small problems instantiating widgets programmatically, because the docs only really describe how to do it with xaml. But what I did was find what widgets I want to use using the docs, then look at the constructors and methods for those classes in my IDE.

In case it helps, Microsoft's documentation for WPF is pretty helpful actually when trying to figure out the widgets. You have to cross check some stuff though.

**Comment by djeikyb** [comment]
Fwiw @Dareka826 published a really helpful example here: https://github.com/Dareka826/Avalonia_no_XAML/blob/master/Program.cs

```cs
using Avalonia;
using Avalonia.Controls;

class MainClass {
    public static void Main(string[] args) {
        AppBuilder
            .Configure<Application>()
            .UsePlatformDetect()
            .Start(AppMain, args);
    }

    public static void AppMain(Application app, string[] args) {
        // Application needs a theme to render window content
        app.Styles.Add(new Avalonia.Themes.Simple.SimpleTheme());
        app.RequestedThemeVariant = Avalonia.Styling.ThemeVariant.Default; // Default, Dark, Light

        // Create window
        var win = new Window();
        win.Title = "Avalonia no XAML";
        win.Width = 800;
        win.Height = 600;

        var text = new Label();
        win.Content = text;

        text.Content = "Hello from C#!";
        text.HorizontalAlignment = Avalonia.Layout.HorizontalAlignment.Center;
        text.VerticalAlignment = Avalonia.Layout.VerticalAlignment.Center;
        text.FontSize = 72;

        win.Show();
        app.Run(win);
    }
}
```

**Reply by tseguine** [reply]
I think I based my code on this example originally (or something very similar to it). The issue I  was trying to solve is that I needed to use it in a scenario that didn't really make sense to have an AppMain function. Specifically I was only using it for a transient splash screen in one code path, and a simple dialog in another. In any case I wanted as little code as necessary. I think my splash screen ended up being shorter and clearer than this example code.

The reason why I originally went on a rant, is that the docs seem to imply everywhere that this library is mostly Windows.Forms compatible.

And in windows forms I could have done what I wanted (at least in terms of shortest path to hello world) more or less with the single line. It's been a while, but If I remember correctly, you just `new Window()` and then call show on it. Which is I think the first thing anyone would think to do.

My issue was that going from the obvious thing to code that actually works was way bigger of an ordeal than seemed necessary.

**Reply by djeikyb** [reply]
Makes sense. I came to this issue trying to draw a grid in the code behind instead of xaml. I prolly won't ship this implementation, but it's an interesting practice problem.

It was incredibly clarifying to understand (or relearn? I feel like I knew this and just forgot) that in the code behind of a control, after calling `InitializeComponent()`, you can set the content property like `this.Content = grid` to whatever you want. Setting it to a control like Grid or StackPanel is interesting because they let you add multiple children.

Not sure how to work all this into a doc, but here's the code version of the previous paragraph:

```cs
// avalonia version 11.0.11
using Avalonia.Controls;

namespace Example;

public partial class MainWindow : Window
{
    public MainWindow()
    {
        InitializeComponent();

        var grid = new Grid();
        grid.ShowGridLines = true;
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        grid.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });

        var text1 = SomeLabel();
        var text2 = SomeLabel();

        Grid.SetColumn(text1, 0);
        Grid.SetColumn(text2, 1);

        grid.Children.Add(text1);
        grid.Children.Add(text2);

        this.Content = grid;
    }

    private static int _i = 0;
    private static Label SomeLabel()
    {
        var text = new Label();
        text.Content = $"halooooo {_i++}";
        text.HorizontalAlignment = Avalonia.Layout.HorizontalAlignment.Center;
        text.VerticalAlignment = Avalonia.Layout.VerticalAlignment.Center;
        text.FontSize = 72;
        return text;
    }
}
```

**Reply by tseguine** [reply]
In case it helps, Microsoft's documentation for WPF is pretty helpful actually when trying to figure out the widgets. You have to cross check some stuff though.

**Reply by stevemonaco** [reply]
You can simplify the `ColumnDefinitions` if you don't mind possibly taking a runtime exception if you mess up the syntax:

```cs
grid.ColumnDefinitions = ColumnDefinitions.Parse("*, *");
```

Either way, if the main goal is to have a minimal crossplatform window, then you're probably better off using GLFW .NET bindings or such. The time and effort spent creating XAML wasn't for the purpose of having people learn more things. As such, no-XAML approaches aren't really represented in the docs though that's equally because of time priorities.

**Comment by stevensrmiller** [comment]
I've written a set of no-XAML examples for those who are interested. There's a recommended order for helping to learn how they work, with comments in the code (the comments assume the reader is following the recommended order, so they don't repeat what's been said before).

I've looked at Avalonia.Declarative.Markup, but haven't been able to understand it yet. If anyone knows of documentation on that, please post a link or let me know. It does look promising.

My repo of no-XAML examples is at [my GitHub page](https://github.com/stevensrmiller/AvaloniaWithoutXAML)

**Comment by bmitc** [comment]
I get the sense from the developers that this is considered an advanced feature and not worth it to document. That is fine, but this feature is prominently advertised on the home page.

![Screenshot_20241011_100755_Chrome](https://github.com/user-attachments/assets/29bc3ffe-e5dd-476a-8a61-3897327e08ad)

This is effectively my first foray into Avalonia. I am coming from Qt Quick with QML and Python as I would prefer building UIs with F# or at worst C# but not necessarily XAML. That this feature was advertised *and* F# was mentioned (!) on the homepage made me very excited and sold me immediately. However, I also found nothing in the documentation even alluding to this, which unsold me, as it makes me worry about its overall support, especially with third-party controls.

I would suggest that if it is not a priority to document this, then it would be best not to advertise it on the homepage.

**Reply by MikeCodesDotNET** [reply]
https://funcui.avaloniaui.net/

**Reply by bmitc** [reply]
I'm hesitant to use that for the similar reasons alluded to above.

For example, how well is FuncUI supported? It's clearly a non-core solution. Will it work with third-party controls, like charts from Syncfusion or SciChart? Does it even support all of the functionality in Avalonia, such as drag and drop?

**Reply by stevensrmiller** [reply]
You could have a look at my no-XAML repo: https://github.com/stevensrmiller/AvaloniaWithoutXAML

I see you said you'd prefer F#, but you'd tolerate C#, which is what my examples all use. They're all pure Avalonia with no other product or library required (other than .NET itself, of course).

The Avalonia devs have spoken to the lack of documentation here and there. The short of it is that they are a small team working hard to pay their bills. The fact that this library is available as FOSS is kind of amazing. It's magnificently platform-agnostic. I use it to teach media coding to my university students, about have of whom use Windows and half use Mac. I have yet to run into a single issue arising from any differences between how Avalonia works on either.

Yes, the lack of documentation on the no-XAML approach is frustrating at times. But if one is persistent, a mix of looking at XAML for keywords, reading what doc there is, Googling, and looking at the Avalonia code itself will always get you though.

**Reply by MikeCodesDotNET** [reply]
I'm genuinely sorry that we don't have the non-XAML options documented. As @stevensrmiller mentioned, we're a small team doing our best to improve things.

Regarding documentation, two years ago, we invested a substantial sum (mid five-figures) in hiring someone on a short-term contract to improve our docs. While this did help improve things, we recognise we've got lots more to do. We funded this work through our commercial activities, as community donations didn't come close to covering the cost. 

For over a year, we've asked our community for GitHub donations to help fund a part-time documentation writer. Currently, we receive about €400 monthly from our community of hundreds of thousands of developers. **If every user donated just €1 a month, we'd have enough to grow our team from 11 to more than 70.** 

While I'd love for us to invest more in documentation, it's just not feasible to keep spending company funds on it when we need to focus on generating revenue (by building products and services). We're constantly trying to balance these priorities to keep our platform growing and sustainable.

**Reply by stevensrmiller** [reply]
This problem seems to plague the creators of lots of popular FOSS products. You become victims of your own success, with people expecting things without making contributions of their own. I really want to commend the Avalonia devs for _not_ falling prey to the defense mechanisms a lot of others use in similar situations (i.e., becoming evasive; dismissing complaints as being evidence of ignorance; telling users they can fork off; simple whining). This library is one of the most mature, fulsome, useable FOSS products I've ever seen. It easily surpasses the usefulness and design quality of a lot of software that people pay for all the time. Somehow, the idea that the people who create such a thing expect also to make a living seems hard for the world to comprehend. FWIW, some of us do.

**Reply by ztl8702** [reply]
Just want to add a data point. I've been using Avalonia for a few months on a daily basis. I did not have WPF experience but it has been very smooth for me. Documentation has not been a big issue for me.

Whenever I am stuck, I almost always got unstuck by searching cs.github.com for examples or 
asking ChatGPT. I rarely needed to raise an issue or ask in chat. I suspect there are a lot of silent happy users. :)

Documentation is nice. But the risk with writing documentation is that they are also a liability since they can get outdated. Nothing is free. The risk is small for users but for the team the opportunity cost could be huge. The proposition that better documentation necessarily will bring more users and therefore more revenue, is pretty hard to justify IMO.

I think the team is doing the right thing by NOT over-investing in onboarding documentation. That is for later in the adoption curve. I see that Avalonia is wisely avoiding chasing "majority adoption" too early. Keep it your way. 👍

**Reply by bmitc** [reply]
I'm unsubscribing from this noisy discussion (happening in other threads), but I just wanted to clarify. It was the advertisement of F# and the non-XAML approach that was advertised on the front page that set the expectation. I came in with no expectations, nor do I expect anything of anyone or anything providing something to me for free.

My point was: don't prominently advertise a feature if it isn't also prominently documented and supported. If the project cannot sustainably do so, then why bother advertising the feature on the home page? Such an approach seems to welcome such discussions as that found here.

**Reply by MikeCodesDotNET** [reply]
@bmitc, Avalonia does support both C# and F#, giving developers more options for how they build apps. A lot of people still assume Avalonia is XAML-only, which limits how the framework is perceived. The goal of the website is to make it clear that it’s possible. 

Regardless of if it should be included on the site or not, we’re actively improving documentation and developer experience across the board. That includes ongoing work and efforts to hire a dedicated technical writer. We’ve already invested a significant amount into this and will continue to reinvest as revenue allows.

We know how important good documentation is, and we’re working hard to improve things. Community feedback and involvement are essential parts of that process. Whether through contributions, funding, or just sharing insight, everyone can help improve Avalonia.
